float32[3] ag_3d